src/Frame.cpp

```diff
@@ -1,227 +1,493 @@
+
+   //Check if total cells is not negative
+   if (totalCells <= 0)
+   {
+      throw error(
+         "Negative number of cells"
+      );
+      return;
+   }
+
+   //Limit number of cells in the box (to avoid memory issues)
+   if (totalCells > 40000000000)
+   {
+      throw error(
+         "Exceeded maximum cells"
+      );
+      return;
+   }
+
+   //Sort atoms in cells
+   std::vector<std::vector<int>> cells;
+   std::vector<int> numInCell;
+   std::vector<int> maxInCell;
+
+   cells.resize(totalCells);
+   numInCell.resize(totalCells);
+   maxInCell.resize(totalCells);
+
+   for (int i = 0; i < numAtoms; i++) {
+      int atomXCellNum, atomYCellNum, atomZCellNum, atomCellIndex, numOfAtomsInCell;
+
+      //Compute atom's cell index
+      Vector3D atomCoord = positions_[i];
+      atomXCellNum = (int)((atomCoord[0] - min[0]) * inversedCutoff);
+      atomYCellNum = (int)((atomCoord[1] - min[1]) * inversedCutoff);
+      atomZCellNum = (int)((atomCoord[2] - min[2]) * inversedCutoff);
+
+      atomCellIndex = atomZCellNum * xyTotalCells + atomYCellNum * xCells + atomXCellNum;
+
+      //Grow cell
+      numOfAtomsInCell = numInCell[atomCellIndex];
+      if (numOfAtomsInCell == maxInCell[atomCellIndex]) {
+         cells[atomCellIndex].resize(numOfAtomsInCell + 4);
+         maxInCell[atomCellIndex] += 4;
+      }
+
+      //Store atom index in appropriate cell
+      cells[atomCellIndex][numOfAtomsInCell] = i;
+      numInCell[atomCellIndex]++;
+   }
+
+   maxInCell.clear();
+
+   //Create the neighbor list 
+   std::vector<std::vector<int>> nbList;
+   nbList.resize(totalCells);
+   createNeighborList(nbList, xCells, yCells, zCells);
+
+   //Max number of 12 bonds per atom
+   int maxNumBonds = numAtoms * 12;
+   std::vector<int> bondResults;
+   int bondsCount = generateBonds(bondResults, totalCells, cells, numInCell, nbList, maxNumBonds, cutoff);
+
+   cells.clear();
+   nbList.clear();
+   numInCell.clear();
+
+   if (bondsCount == 0 || bondsCount > maxNumBonds)
+   {
+      if (bondsCount == 0)
+      {
+         throw error(
+            "no bonds found"
+         );
+      }
+      else if (bondsCount > maxNumBonds)
+      {
+         throw error(
+            "exceeded max number of bonds"
+         );
+      }
+   }
+
+   // Go through the bond results vector to add bonds to connectivity
+   for (int bond = 0; bond < bondResults.size() / 2; bond++) {
+
+      int atom1Index = bondResults[bond * 2];
+      int atom2Index = bondResults[(bond * 2) + 1];
+
+      Atom atom1 = topology_[atom1Index];
+      Atom atom2 = topology_[atom2Index];
+
+      //Prevent hydrogen atoms from bonding with each other
+      std::string atom1Type = atom1.type();
+      std::string atom2Type = atom2.type();
+      if (atom1Type != "H" || atom2Type != "H") {
+         // Add a bond, bondorder defaults to 1, bond type to -1
+         add_bond(atom1Index, atom2Index);
+      }
+   }
+
+   bondResults.clear();
```

Comment:
no need to clear, this is a local variable so memory willl be released when the scope exit.